internal/controller/gitrepository_controller.go

```diff
@@ -444,6 +446,97 @@ func (r *GitRepositoryReconciler) reconcileStorage(ctx context.Context, sp *patc
 	return sreconcile.ResultSuccess, nil
 }
 
+// Interface to configure gitclient with custom TLS options
+// used for mTLS authentication.
+type GitClientConfigurer interface {
+	ConfigureGitClient(ctx context.Context, obj *sourcev1.GitRepository)
+	IsValid() bool
+}
+
+type GitClientHttpConfigurer struct {
+	SSLCertificateData map[string][]byte
+	Valid              bool
+	DefaultTransport   transport.Transport
+	MTLSTransport      transport.Transport
+}
+
+func (c *GitClientHttpConfigurer) IsValid() bool {
+	return c.Valid
+}
+
+func (c *GitClientHttpConfigurer) SetValid() {
+	c.Valid = true
+}
+
+func (c *GitClientHttpConfigurer) SetInvalid() {
+	c.Valid = false
+}
+
+// Ensures the secret contains the valid keys for the certificate & private key.
+func (r *GitRepositoryReconciler) isCertificateDataValid(sslCertificateData map[string][]byte) bool {
+	certBytes, keyBytes := sslCertificateData["tls.crt"], sslCertificateData["tls.key"]
+	// Validate that both the certificate and key data are present
+	return len(certBytes) > 0 && len(keyBytes) > 0
+}
+
+// Configure the gitclient with a tlsconfig containing the certifcates & CA specified in the spec.secretRef.name secret.
+func (h *GitClientHttpConfigurer) ConfigureGitClient(ctx context.Context, obj *sourcev1.GitRepository) {
+
+	if obj.Spec.SecretRef != nil {
+		sslCertificate := &corev1.Secret{
+			Data: h.SSLCertificateData,
+		}
+		tlsConfig, _, err := tls.KubeTLSClientConfigFromSecret(*sslCertificate, "")
+		if err != nil {
+			fmt.Println("Error generating TLS config:", err)
+			return
+		}
+
+		var transportHttp transport.Transport
+		if tlsConfig != nil {
+
+			transportHttp, err = HttpTransportwithCustomCerts(tlsConfig, ctx)
+			if err != nil {
+				fmt.Println("Error setting up transport:", err)
+				return
+			}
+		} else {
+			transportHttp = h.DefaultTransport
+		}
+
+		gitclient.InstallProtocol("https", transportHttp)
```

Comment:
We can't alter the global client as this would impact all GitRepositories. If you have GitHub repo and some mTLS Git server, then GitHub will fail.